pkg/ingester/ingester.go

```diff
@@ -972,82 +1090,66 @@ func (i *Ingester) pushSamplesToAppender(userID string, timeseries []mimirpb.Pre
 				continue
+			}
 
-			case storage.ErrOutOfOrderSample:
-				stats.sampleOutOfOrderCount++
-				updateFirstPartial(func() error {
-					if ephemeral {
-						return newEphemeralIngestErrSampleOutOfOrder(model.Time(s.TimestampMs), ts.Labels)
-					}
-					return newIngestErrSampleOutOfOrder(model.Time(s.TimestampMs), ts.Labels)
-				})
-				continue
+			// Otherwise, return a 500.
+			return wrapWithUser(err, userID)
+		}
 
-			case storage.ErrTooOldSample:
-				stats.sampleTooOldCount++
-				updateFirstPartial(func() error {
-					// OOO is not enabled for ephemeral storage, so we can't get this error.
-					return newIngestErrSampleTimestampTooOldOOOEnabled(model.Time(s.TimestampMs), ts.Labels, outOfOrderWindow)
-				})
-				continue
+		if len(ts.Histograms) > 0 {
+			if i.limits.AcceptNativeHistograms(userID) {
+				for _, h := range ts.Histograms {
+					var (
+						err error
+						ih  *histogram.Histogram
+						fh  *histogram.FloatHistogram
+					)
 
-			case storage.ErrDuplicateSampleForTimestamp:
-				stats.newValueForTimestampCount++
-				updateFirstPartial(func() error {
-					if ephemeral {
-						return newEphemeralIngestErrSampleDuplicateTimestamp(model.Time(s.TimestampMs), ts.Labels)
+					if h.IsFloatHistogram() {
+						fh = mimirpb.FromHistogramProtoToFloatHistogram(&h)
+					} else {
+						ih = mimirpb.FromHistogramProtoToHistogram(&h)
+					}
+					if ref != 0 {
+						if _, err = app.AppendHistogram(ref, copiedLabels, h.Timestamp, ih, fh); err == nil {
+							stats.histogramStats.succeededSamplesCount++
+							continue
+						}
+					} else {
+						copiedLabels = mimirpb.FromLabelAdaptersToLabelsWithCopy(ts.Labels)
+						if ref, err = app.AppendHistogram(0, copiedLabels, h.Timestamp, ih, fh); err == nil {
+							stats.histogramStats.succeededSamplesCount++
+							continue
+						}
 					}
-					return newIngestErrSampleDuplicateTimestamp(model.Time(s.TimestampMs), ts.Labels)
-				})
-				continue
 
-			case errMaxSeriesPerUserLimitExceeded, errMaxEphemeralSeriesPerUserLimitExceeded: // we have special error for this, as we want different help message from FormatError.
-				stats.perUserSeriesLimitCount++
-				updateFirstPartial(func() error {
-					if ephemeral {
-						return makeLimitError(i.limiter.FormatError(userID, cause))
+					stats.histogramStats.failedSamplesCount++
+
+					if handleAppendError(err, h.Timestamp, ts.Labels, copiedLabels, &stats.histogramStats) {
+						continue
 					}
-					return makeLimitError(i.limiter.FormatError(userID, cause))
-				})
-				continue
 
-			case errMaxSeriesPerMetricLimitExceeded:
-				stats.perMetricSeriesLimitCount++
-				updateFirstPartial(func() error {
-					// Ephemeral storage doesn't have this limit.
-					return makeMetricLimitError(copiedLabels, i.limiter.FormatError(userID, cause))
-				})
-				continue
+					return wrapWithUser(err, userID)
+				}
+			} else { // ignore histograms and increase counter
+				stats.histogramStats.failedSamplesCount++
```

Comment:
How do we benefit from this stat? Isn't it misleading? Because it technically did not fail; it was just not enabled for this tenant. This can lead to false alerts if it was based on this.

Reply:
that's fair

Reply:
done